File: src/com/shubh/java/MergeSort.java
package com.shubh.java;

import java.util.Arrays;

public class MergeSort implements SortingAlgo{

    @Override
    public int[] sort(int[] input) {
        mergeSort(input);
        System.out.println( Arrays.toString(input));
        return input;
    }

    void mergeSort(int[] arr) {
        int n = arr.length;
        if(n == 1) return;

        int nL = n/2;
        int nR = n - nL;
        int[] left = new int[nL];
        int[] right = new int[nR];

        int i = 0,k = 0;
        while( i < nL) {
            left[i++] = arr[k++];
        }
        i = 0;
        while( i < nR) {
            right[i++] = arr[k++];
        }

        mergeSort(left);
        mergeSort(right);
        merge(arr, left, right);
    }

   private void merge(int[] arr, int[] left, int[] right) {
        int nL = left.length;
        int nR = right.length;

        int i = 0,j = 0, k = 0;

        while (i < nL && j < nR) {
            if(left[i] <= right[j]) {
                arr[k++] = left[i++];
            } else {
                arr[k++] = right[j++];
            }
        }

        while (i < nL){
            arr[k++] = left[i++];
        }

        while (j < nR){
            arr[k++] = right[j++];
        }
   }
}

File: src/com/shubh/kotlin/MergeSort.kt
package com.shubh.kotlin;

class MergeSort(arr: Array<Int>, l: Int = 0, r: Int = arr.size-1) {

    init {
        if(l < r) {
            var m = l + (r-l)/2
            MergeSort(arr, l , m)
            MergeSort(arr, m+1 , r)
            merge(arr,l,m,r)
        }
        arr.print()
    }

    private fun merge(arr: Array<Int>, l: Int, m: Int, r: Int) : Int {
        var n1: Int = m-l+1
        var n2: Int = r-m

        var left = Array<Int>(n1){0};
        var right = Array<Int>(n2){0};

        left.forEachIndexed { index, _ ->
            left[index] = arr[index]
        }

        right.forEachIndexed { index, _ ->
            right[index] = arr[index+m]
        }

        var i = 0;
        var j = 0;
        var k = 0;

        while(i < n1 && j < n2) {
            if(left[i] < right[j]) {
                arr[k++] = left[i++]
            } else {
                arr[k++] = right[j++]
            }
        }

        while (i < n1) {
            arr[k++] = left[i++]
        }

        while (j < n2) {
            arr[k++] = right[j++]
        }


        return 0;
    }

}

